### PR TITLE
pkgs/jni: Support Throwing Java Exceptions in Interface Impl.

### PR DESCRIPTION
**Pull request by Anikate-De**
Items in this PR:
 - [x] #1209 
 - [x] Add `throw Java exceptions on the same thread` test case to pkgs/jnigen/test/simple_package_test 

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by Anikate-De** [comment]
> To regenerate the bindings for jni, run the script in `tool/generate_ffi_bindings.dart`. And after changing the code generation in jnigen, you can generate all the bindings again using `tool/regenerate_all_bindings.dart` (in jnigen).

I haven't used the jnigen regeneration script yet as I was using a very particular use case of `ok_http` and manually regenerating the bindings when needed.

I will soon shift to a more general example of throwing exceptions in interface implementation to test this PR overall, and I will use `tool/regenerate_all_bindings.dart` then (thanks :))

**Comment by Anikate-De** [comment]
Thanks for all the help!

I temporarily edited the `kotlin_plugin` example within `jnigen` to demonstrate (and test) that this works. It has a very crude interface `Divider`. And a class `DividerUser` that catches any `java.lang.ArithmeticException`s thrown by the former.
We then deliberately throw the same, and expect a `0` as a result, since it will be caught in the Kotlin code.

No more `java.lang.reflect.UndeclaredThrowableException`!

If you feel the logical contents of this PR are satisfactory, then I can revert the last commit (Editing the jnigen example), and then get it ready for review.

File: pkgs/jnigen/lib/src/bindings/dart_generator.dart [hunk 1]
```diff
@@ -413,7 +413,10 @@ class $name$typeParamsDef extends $superName {
       }
       s.write('''
     } catch (e) {
-      return $_protectedExtension.newDartException(e.toString());
+      if (e is jni.JObject && jni.Jni.env.IsInstanceOf(e.reference.pointer, jThrowableClass.reference.pointer)) {
```
Comment (HosseinYousefi) [inline pkgs/jnigen/lib/src/bindings/dart_generator.dart:416]: Maybe this logic could reside in `ProtectedJniExtensions` as well so we can just generate:

```dart
catch(e) {
  return $_protectedExtension.newDartException(e);
}
```

And the method does the checks and `toString` and all.

**Comment by HosseinYousefi** [comment]
Great job, @Anikate-De!

Yes, revert the temp commit and add the test to `test/simple_package` in jnigen. We currently cram all the usecases there. There are other tests for the interface implementation there as well.

File: pkgs/jnigen/lib/src/bindings/dart_generator.dart (hunk 1 above)
Comment (Anikate-De) [reply pkgs/jnigen/lib/src/bindings/dart_generator.dart]: That sounds good! And, done ✅

File: pkgs/jnigen/test/simple_package_test/runtime_test_registrant.dart [hunk 2]
```diff
@@ -687,6 +687,29 @@ void registerTests(String groupName, TestRunnerCallback test) {
...
       }
     });
+
+    group('throw Java exceptions', () {
+      test('StringConverter.implement', () async {
+        final stringConverter = StringConverter.implement($StringConverterImpl(
+          parseToInt: (s) {
+            final firstChar = s.toDartString().substring(0, 1);
+            final value = int.tryParse(firstChar);
+            if (value == null || value > 5) {
```
Comment (HosseinYousefi) [inline pkgs/jnigen/test/simple_package_test/runtime_test_registrant.dart:697]: nit: This is a weird logic, why should `parseToInt` fail if `value > 5`? Why do we only parse the first digit?

I think it's better to just go with `tryParse` and only throw when that's null. Your test could be trying to convert `foo` (and failing, so -1) and `700` (and succeeding). It's easier to understand the test this way and it achieves the same thing.

**Review by HosseinYousefi — CHANGES_REQUESTED** [review]

**Comment by HosseinYousefi** [comment]
Also please add a changelog entry to jnigen about this.

File: pkgs/jni/lib/src/jni.dart [hunk 3]
```diff
@@ -7,12 +7,12 @@ import 'dart:io';
 import 'dart:isolate';
 
 import 'package:ffi/ffi.dart';
+import 'package:jni/internal_helpers_for_jnigen.dart' show jThrowableClass;
```
[the remark this answers is not in the document]
Comment (Anikate-De) [reply pkgs/jni/lib/src/jni.dart]: Done ✅

File: pkgs/jnigen/test/simple_package_test/java/com/github/dart_lang/jnigen/interfaces/StringConverterConsumer.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+package com.github.dart_lang.jnigen.interfaces;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.FutureTask;
+
+public class StringConverterConsumer {
```
[the remark this answers is not in the document]
Comment (Anikate-De) [reply pkgs/jnigen/test/simple_package_test/java/com/github/dart_lang/jnigen/interfaces/StringConverterConsumer.java:7]: Done ✅

File: pkgs/jnigen/test/simple_package_test/runtime_test_registrant.dart [hunk 5]
```diff
@@ -687,6 +687,29 @@ void registerTests(String groupName, TestRunnerCallback test) {
...
+            }
+
+            return value;
+          },
+        ));
+
+        final testString = '700'.toJString();
+        final result = StringConverterConsumer.consumeOnSameThread(
+            stringConverter, testString);
+        expect(result, -1);
```
[the remark this answers is not in the document]
Comment (Anikate-De) [reply pkgs/jnigen/test/simple_package_test/runtime_test_registrant.dart:712]: Done ✅

File: pkgs/jnigen/test/simple_package_test/runtime_test_registrant.dart (hunk 2 above)
Comment (Anikate-De) [reply pkgs/jnigen/test/simple_package_test/runtime_test_registrant.dart]: Done ✅ 

I was actually exploring throwing multiple exceptions and only handling a select few in the Java code, but that turned out to be troublesome and I had to remove the other custom exceptions.

**Comment by Anikate-De** [comment]
Bumped up version to `0.9.3`

**Comment by Anikate-De** [comment]
There is a problem with throwing exceptions when the interface is consumed on a separate thread. (the reason why `consumeOnAnotherThread` is missing in the test case)

Quite like `ok_http`, if we throw the exception in that case, we run into a deadlock. For `ok_http`, it is default behaviour to terminate the connection with a `java.net.SocketTimeoutException`, but here, **it pretty much leads to a forever wait cycle.**

How could we solve this?

File: pkgs/jni/lib/src/jni.dart [hunk 6]
```diff
@@ -221,6 +220,8 @@ typedef _SetJniGettersDartType = void Function(Pointer<Void>, Pointer<Void>);
 
 /// Extensions for use by `jnigen` generated code.
 extension ProtectedJniExtensions on Jni {
+  static final jThrowableClass = JClass.forName('java/lang/Throwable');
```
Comment (HosseinYousefi) [inline pkgs/jni/lib/src/jni.dart:223]: nit: make private

**Review by HosseinYousefi — CHANGES_REQUESTED** [review]

**Review by HosseinYousefi — APPROVED** [review]

**Comment by HosseinYousefi** [comment]
Added the test for the `Future` case. 

Thanks @Anikate-De!

**Pull request merged**